# Peer-to-Peer Credit Risk Analysis

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. This program uses various techniques to train and evaluate models with imbalanced classes. It uses a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

In [56]:
# Imported the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from imblearn.over_sampling import RandomOverSampler
from imblearn.metrics import classification_report_imbalanced

import warnings
warnings.filterwarnings('ignore')

In [57]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_df = pd.read_csv(Path('Resources/lending_data.csv'))

# Review the DataFrame
lending_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [23]:
# Separated the data into labels:(y) and features:(X)
y = lending_df['loan_status']

X = lending_df.drop(columns=['loan_status'])

In [24]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [25]:
# Review the X variable DataFrame
X[:5]

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

In [58]:
# Checked the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

In [64]:
# Split the data using train_test_split
# Assigned a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [65]:
# Instantiated the Logistic Regression model
# Assigned a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fitted the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

In [66]:
# Made a prediction using the testing data
model_pred = lr_model.predict(X_test)

In [67]:
# Printed the balanced_accuracy score of the model
baso = balanced_accuracy_score(y_test, model_pred)
print(baso)

0.9520479254722232


In [68]:
# Generated a confusion matrix for the model
test_matrix = confusion_matrix(y_test, model_pred)
print(test_matrix)

[[18663   102]
 [   56   563]]


In [69]:
# Printed the classification report for the model
test_report = classification_report_imbalanced(y_test, model_pred)
print(test_report)

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.91      1.00      0.95      0.91     18765
          1       0.85      0.91      0.99      0.88      0.95      0.90       619

avg / total       0.99      0.99      0.91      0.99      0.95      0.91     19384



**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
    This Logistic Regression model has a balanced accuracy score of about 95%. The recall amount for `1`(high-risk loan) is 91%, which means that 91% of the high risk loans were identified using the model. The precision of `1`(high-risk loan) is 85%, which means that in this model there is a 15% rate of false positives when identifying high risk loans. 

   This model is able to predict `0`(healthy loan) extreamly well with the recall at 99% which means that about 99% of the healthy loans were identified and had a percision of 100%, which means that in this model there is a 0% rate of false positives when identifying healthy loans.

In [70]:
# Instantiated the RandomOverSampler model
# # Assigned a random_state parameter of 1 to the model
random_oversampler = RandomOverSampler(random_state=1)

# Fitted the original training data to the random_oversampler model
X_resampled, y_resampled = random_oversampler.fit_resample(X_train, y_train)

In [71]:
# Counted the distinct values of the resampled labels data
y_resampled.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

In [72]:
# Instantiated the Logistic Regression model
# Assigned a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)

# Fitted the model using the resampled training data
lr_resampled_model = model.fit(X_resampled, y_resampled)

# Made a prediction using the testing data
y_resampled_pred = lr_resampled_model.predict(X_test)

In [73]:
# Printed the balanced_accuracy score of the model 
basr = balanced_accuracy_score(y_test, y_resampled_pred)
print(basr)

0.9936781215845847


In [76]:
# Generated a confusion matrix for the model
resampled_matrix = confusion_matrix(y_test, y_resampled_pred)
print(resampled_matrix)

[[18649   116]
 [    4   615]]


In [77]:
# Printed the classification report for the model
resampled_report = classification_report_imbalanced(y_test, y_resampled_pred)
print(resampled_report)

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.99      1.00      0.99      0.99     18765
          1       0.84      0.99      0.99      0.91      0.99      0.99       619

avg / total       0.99      0.99      0.99      0.99      0.99      0.99     19384



**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** This Oversampled Logistic Regression model has a balanced accuracy score of about 99%. The recall amount for `1`(high-risk loan) is 99%, which means that 99% of the high risk loans were identified using the oversampled model. The precision of `1`(high-risk loan) is 84%, which means that in this model there is a 16% rate of false positives when identifying high risk loans. 

   This model is able to predict `0`(healthy loan) extreamly well again with the recall at 99% which means that about 99% of the healthy loans were identified and had a percision of 100%, which means that in this model there is a 0% rate of false positives when identifying healthy loans.
   
   Comparing the original model against the oversampled model we can see that the oversampled model is able to identify high-risk loans much better than the orginal, while only losing 1% in precision. Given this information it is better to use the oversampled model when utilizing this set of data for predictions.